pkg/controller/cnr/cnr.go

```diff
@@ -0,0 +1,301 @@
+		client:       client,
+		cnrSyncQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), cnrControllerName),
+		podTimeMap:   sync.Map{},
+	}
+
+	// init cnr informer
+	cnrInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    cnrController.addCNREventHandler,
+		UpdateFunc: cnrController.updateCNREventHandler,
+	})
+	// init cnr lister
+	cnrController.cnrLister = cnrInformer.Lister()
+	// init cnr synced
+	cnrController.cnrListerSynced = cnrInformer.Informer().HasSynced
+
+	// init pod informer
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: cnrController.updatePodEventHandler,
+	})
+	// init pod lister
+	cnrController.podLister = podInformer.Lister()
+	// init pod synced
+	cnrController.podListerSynced = podInformer.Informer().HasSynced
+
+	if metricsEmitter == nil {
+		// if metricsEmitter is nil, use dummy metrics
+		cnrController.metricsEmitter = metrics.DummyMetrics{}
+	} else {
+		// if metricsEmitter is not nil, use metricsEmitter with tags
+		cnrController.metricsEmitter = metricsEmitter.WithTags(cnrControllerName)
+	}
+
+	// init cnr control
+	cnrController.cnrControl = control.DummyCNRControl{}
+	if !genericConf.DryRun {
+		// if dryRun is false, use real cnr control
+		cnrController.cnrControl = control.NewCNRControlImpl(client.InternalClient)
+	}
+
+	return cnrController, nil
+}
+
+func (ctrl *CNRController) Run() {
+	defer utilruntime.HandleCrash()
+	defer ctrl.cnrSyncQueue.ShutDown()
+	defer klog.Infof("Shutting down %s controller", cnrControllerName)
+
+	// wait for cnr cache sync
+	if !cache.WaitForCacheSync(ctrl.ctx.Done(), ctrl.cnrListerSynced, ctrl.podListerSynced) {
+		utilruntime.HandleError(fmt.Errorf("unable to sync caches for %s controller", cnrControllerName))
+		return
+	}
+
+	klog.Infof("Caches are synced for %s controller", cnrControllerName)
+	klog.Infof("start %d workers for %s controller", cnrWorkerCount, cnrControllerName)
+
+	for i := 0; i < cnrWorkerCount; i++ {
+		go wait.Until(ctrl.cnrWorker, time.Second, ctrl.ctx.Done())
+	}
+
+	// gc podTimeMap
+	klog.Infof("start gc podTimeMap...")
+	go wait.Until(ctrl.gcPodTimeMap, 1*time.Minute, ctrl.ctx.Done())
+
+	<-ctrl.ctx.Done()
+}
+
+func (ctrl *CNRController) cnrWorker() {
+	for ctrl.processNextCNR() {
+	}
+}
+
+// processNextCNR dequeues items, processes them, and marks them done.
+// It enforces that the sync is never invoked concurrently with the same key.
+func (ctrl *CNRController) processNextCNR() bool {
+	key, quit := ctrl.cnrSyncQueue.Get()
+	if quit {
+		return false
+	}
+	defer ctrl.cnrSyncQueue.Done(key)
+
+	err := ctrl.syncCNR(key.(string))
+	if err == nil {
+		ctrl.cnrSyncQueue.Forget(key)
+		return true
+	}
+
+	// if err is not nil, requeue the key
+	utilruntime.HandleError(fmt.Errorf("sync %s failed with %v", key, err))
+	ctrl.cnrSyncQueue.AddRateLimited(key)
+
+	return true
+}
+
+func (ctrl *CNRController) syncCNR(key string) error {
```

Comment:
the cnr anomaly check no need be triggered by event, it may list all cnrs every 30s to check more properly.

Reply:
I was thinking that the kcnr data anomaly is a small probability event, if the rotation check method is adopted, consider a situation, if the number of nodes and pods is large, then each query needs to traverse these cnrs and their contained Pods, maybe it will affect the performance.

Reply:
another choice is that if there is kcnr with anomaly, re-enqueue it use AddAfter func with 30s duration.